File: ShadowFrameworkLite20/src/shadow/image/SFDrawnRenderedTexture.java
package shadow.image;

import shadow.pipeline.SFPipeline;
import shadow.renderer.SFNode;
import shadow.renderer.SFRenderer;

/**
 * @author Alessandro Martinelli
 */
public class SFDrawnRenderedTexture implements SFRenderedTexturesSet{

	private boolean initialized=false;
	private SFPipelineTexture[] textures;
	private boolean useDefaultDepthBuffer=false;
	private boolean useDefaultStencilBuffer=false;
	private int depthBuffer=-1;
	private int stencilBuffer=-1;
	private SFRenderer renderer=new SFRenderer();
	private SFNode node;
	SFRenderedTexture renderedTexture=new SFRenderedTexture();
	
	@Override
	public SFPipelineTexture getTexture(int index) {
		return textures[index];
	}
	
	@Override
	public int getTextureSize() {
		return textures.length;
	}

	@Override
	public void init() {
		if(renderer==null || node==null)
			return;
		
		if(!initialized){
			
			for (int i = 0; i < textures.length; i++) {
				if(i!=depthBuffer && i!=stencilBuffer){
					SFPipelineTexture data=textures[i];
					textures[i]=SFPipeline.getSfTexturePipeline().getRenderedTextureFactory().generateTextureBuffer(data.getWidth(), data.getHeight(), 
							data.getFormat(),  data.getFilters(),	data.getWrapS(), data.getWrapT());
					renderedTexture.addColorData(textures[i]);
				}
			}
			
			if(depthBuffer>=0){
				SFPipelineTexture data=textures[depthBuffer];
				textures[depthBuffer]=SFPipeline.getSfTexturePipeline().getRenderedTextureFactory().generateTextureBuffer(data.getWidth(), data.getHeight(), 
						data.getFormat(),  data.getFilters(),	data.getWrapS(), data.getWrapT());
				renderedTexture.setDepthBuffer(textures[depthBuffer]);
			}else if(useDefaultDepthBuffer){
				SFPipelineTexture data=textures[0];
				SFBufferData bufferData=SFPipeline.getSfTexturePipeline().getRenderedTextureFactory().
					generatePlainBuffer(data.getWidth(), data.getHeight());
				renderedTexture.setDepthBuffer(bufferData);
			}
			
			if(stencilBuffer>=0){
				SFPipelineTexture data=textures[stencilBuffer];
				textures[stencilBuffer]=SFPipeline.getSfTexturePipeline().getRenderedTextureFactory().generateTextureBuffer(data.getWidth(), data.getHeight(), 
						data.getFormat(),  data.getFilters(),	data.getWrapS(), data.getWrapT());
				renderedTexture.setStencilBuffer(textures[stencilBuffer]);
			}else if(useDefaultStencilBuffer){
				SFPipelineTexture data=textures[0];
				SFBufferData bufferData=SFPipeline.getSfTexturePipeline().getRenderedTextureFactory().
					generatePlainBuffer(data.getWidth(), data.getHeight());
				renderedTexture.setStencilBuffer(bufferData);
			}
			
			SFPipeline.getSfTexturePipeline().beginNewRenderedTexture(renderedTexture);

				float[] matrix=renderer.getCamera().extractTransform();
				SFPipeline.getSfPipelineGraphics().setupProjection(matrix);
				renderer.render(node);
			
				renderer.getCamera().getF().set3f(-2.0f, 0, 0);
				matrix=renderer.getCamera().extractTransform();
				SFPipeline.getSfPipelineGraphics().setupProjection(matrix);
				renderer.render(node);
			
				renderer.getCamera().getF().set3f(+2.0f, 0, 0);
				matrix=renderer.getCamera().extractTransform();
				SFPipeline.getSfPipelineGraphics().setupProjection(matrix);
				renderer.render(node);
			
				renderer.getCamera().getF().set3f(0, +2.0f, 0);
				matrix=renderer.getCamera().extractTransform();
				SFPipeline.getSfPipelineGraphics().setupProjection(matrix);
				renderer.render(node);
			
				renderer.getCamera().getF().set3f(0, -2.0f, 0);
				matrix=renderer.getCamera().extractTransform();
				SFPipeline.getSfPipelineGraphics().setupProjection(matrix);
				renderer.render(node);
				
			SFPipeline.getSfTexturePipeline().endRenderedTexture(renderedTexture);
			
			SFPipeline.getSfTexturePipeline().destroyRenderedTexture(renderedTexture);
		}
		initialized=true;
	}
	
	@Override
	public void update() {
		
		
	}
	
	@Override
	public void destroy() {
		if(initialized){
			for (int i = 0; i < textures.length; i++) {
				SFPipeline.getSfTexturePipeline().getRenderedTextureFactory().destroyBuffer(textures[i]);
			}
		}
	}

	public int getDepthBuffer() {
		return depthBuffer;
	}

	public void setDepthBuffer(int depthBuffer) {
		this.depthBuffer = depthBuffer;
	}

	public int getStencilBuffer() {
		return stencilBuffer;
	}

	public void setStencilBuffer(int stencilBuffer) {
		this.stencilBuffer = stencilBuffer;
	}

	public void setTextures(SFPipelineTexture[] textures) {
		this.textures = textures;
	}

	public void setRenderer(SFRenderer renderer) {
		this.renderer = renderer;
	}

	public void setNode(SFNode node) {
		this.node = node;
	}

	public boolean isUseDefaultDepthBuffer() {
		return useDefaultDepthBuffer;
	}

	public void setUseDefaultDepthBuffer(boolean useDefaultDepthBuffer) {
		this.useDefaultDepthBuffer = useDefaultDepthBuffer;
	}

	public boolean isUseDefaultStencilBuffer() {
		return useDefaultStencilBuffer;
	}

	public void setUseDefaultStencilBuffer(boolean useDefaultStencilBuffer) {
		this.useDefaultStencilBuffer = useDefaultStencilBuffer;
	}

}

File: trunk/ShadowFrameworkJS/src/shadow/image/SFDrawnRenderedTexture.js


function SFDrawnRenderedTexture(){
	
	this.initialized=false;
	this.textures=null;
	this.useDefaultDepthBuffer=false;
	this.useDefaultStencilBuffer=false;
	this.depthBuffer=-1;
	this.stencilBuffer=-1;
	this.renderer=new SFRenderer();
	this.node=null;
	this.renderedTexture=new SFRenderedTexture();
}


SFDrawnRenderedTexture.prototype["getTexture"]=function(index){
	return this.textures[index];
};

SFDrawnRenderedTexture.prototype["getTextureSize"]=function(){
	return this.textures.length;
};

SFDrawnRenderedTexture.prototype["init"]=function(){
	if(this.renderer==null || this.node==null)
			return;
		
		if(!this.initialized){
			
			for (var i = 0; i < this.textures.length; i++) {
				if(i!=this.depthBuffer && i!=this.stencilBuffer){
					var data=this.textures[i];
					this.textures[i]=SFPipeline_getSfTexturePipeline().getRenderedTextureFactory().
							generateTextureBuffer(data.getWidth(), data.getHeight(), 
							data.getFormat(),  data.getFilters(),
							data.getWrapS(), data.getWrapT());
					this.renderedTexture.addColorData(this.textures[i]);
				}
			}
			
			
			if(this.depthBuffer>=0){
				var data=this.textures[this.depthBuffer];
				this.textures[depthBuffer]=SFPipeline_getSfTexturePipeline().getRenderedTextureFactory().
						generateTextureBuffer(this.data.getWidth(), this.data.getHeight(), 
						this.data.getFormat(),  this.data.getFilters(),	this.data.getWrapS(), this.data.getWrapT());
				this.renderedTexture.setDepthBuffer(this.textures[this.depthBuffer]);
			}else if(this.useDefaultDepthBuffer){
				var data=this.textures[0];
				var bufferData=SFPipeline_getSfTexturePipeline().getRenderedTextureFactory().
					generatePlainBuffer(this.data.getWidth(), this.data.getHeight());
				this.renderedTexture.setDepthBuffer(bufferData);
			}
			
			if(this.stencilBuffer>=0){
				var data=textures[this.stencilBuffer];
				this.textures[stencilBuffer]=SFPipeline_getSfTexturePipeline().getRenderedTextureFactory().
						generateTextureBuffer(this.data.getWidth(), this.data.getHeight(), 
						this.data.getFormat(),  this.data.getFilters(),	this.data.getWrapS(), this.data.getWrapT());
				this.renderedTexture.setStencilBuffer(this.textures[this.stencilBuffer]);
			}else if(this.useDefaultStencilBuffer){
				var data=this.textures[0];
				var bufferData=SFPipeline_getSfTexturePipeline().getRenderedTextureFactory().
					generatePlainBuffer(this.data.getWidth(), this.data.getHeight());
				this.renderedTexture.setStencilBuffer(this.bufferData);
			}
			
			SFPipeline_getSfTexturePipeline().beginNewRenderedTexture(this.renderedTexture);

				var matrix=this.renderer.getCamera().extractTransform();
				SFPipeline_getSfPipelineGraphics().setupProjection(matrix);
				this.renderer.render(this.node);
			
				this.renderer.getCamera().getF().set3f(new SFVertex3f(-2.0, 0, 0));
				matrix=this.renderer.getCamera().extractTransform();
				SFPipeline_getSfPipelineGraphics().setupProjection(matrix);
				this.renderer.render(this.node);
			
				this.renderer.getCamera().getF().set3f(new SFVertex3f(+2.0, 0, 0));
				matrix=this.renderer.getCamera().extractTransform();
				SFPipeline_getSfPipelineGraphics().setupProjection(matrix);
				this.renderer.render(this.node);
			
				this.renderer.getCamera().getF().set3f(new SFVertex3f(0, +2.0, 0));
				matrix=this.renderer.getCamera().extractTransform();
				SFPipeline_getSfPipelineGraphics().setupProjection(matrix);
				this.renderer.render(this.node);
			
				this.renderer.getCamera().getF().set3f(new SFVertex3f(0, -2.0, 0));
				matrix=this.renderer.getCamera().extractTransform();
				SFPipeline_getSfPipelineGraphics().setupProjection(matrix);
				this.renderer.render(this.node);
				
			SFPipeline_getSfTexturePipeline().endRenderedTexture(this.renderedTexture);
			
			SFPipeline_getSfTexturePipeline().destroyRenderedTexture(this.renderedTexture);
		}
		this.initialized=true;
};

SFDrawnRenderedTexture.prototype["update"]=function(){
	//
};
	
SFDrawnRenderedTexture.prototype["destroy"]=function(){
	if(this.initialized){
		for (var i = 0; i < this.textures.length; i++) {
			SFPipeline_getSfTexturePipeline().getRenderedTextureFactory().destroyBuffer(this.textures[i]);
		}
	}
};


SFDrawnRenderedTexture.prototype["getDepthBuffer"]=function(){
	return this.depthBuffer;
};	
	
SFDrawnRenderedTexture.prototype["setDepthBuffer"]=function(depthBuffer){
		this.depthBuffer = depthBuffer;
};	
	
SFDrawnRenderedTexture.prototype["stencilBuffer"]=function(depthBuffer){
	return this.stencilBuffer;
};	


SFDrawnRenderedTexture.prototype["setStencilBuffer"]=function(stencilBuffer){
		this.stencilBuffer = stencilBuffer;
};	

SFDrawnRenderedTexture.prototype["setTextures"]=function(textures){
		this.textures = textures;
};	


SFDrawnRenderedTexture.prototype["setRenderer"]=function(renderer){
		this.renderer=renderer;
};	

SFDrawnRenderedTexture.prototype["setNode"]=function(node){
		this.node=node;
};	

SFDrawnRenderedTexture.prototype["isUseDefaultDepthBuffer"]=function(){
		return this.useDefaultDepthBuffer;
};

SFDrawnRenderedTexture.prototype["setUseDefaultDepthBuffer"]=function(useDefaultDepthBuffer){
		this.useDefaultDepthBuffer = useDefaultDepthBuffer;
};

SFDrawnRenderedTexture.prototype["isUseDefaultStencilBuffer"]=function(){
		return this.useDefaultStencilBuffer;
};

SFDrawnRenderedTexture.prototype["setUseDefaultStencilBuffer"]=function(useDefaultStencilBuffer){
		this.useDefaultStencilBuffer = useDefaultStencilBuffer;
};

